### PR TITLE
ocamlPackages.buildDunePackage: add support for dune 3 (w/ related additions & updates)

### PR DESCRIPTION
**Pull request by vbgl**
###### Description of changes

Motivated by https://github.com/NixOS/nixpkgs/pull/163766#issuecomment-1066100292 & #164287

This adds a new argument `duneVersion` to `buildDunePackages` that have precedence over `useDune2`. It can be used to choose which version of dune to use.

This PR (as a show case) makes `stog` use dune 3 (it just works).

This adds a few dune-3 related libraries.

This updates dune-related libraries that were already packaged.

cc @AllanBlanchard @marsam

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vbgl** [comment]
ping @sternenseemann, in case he’s interested & did not notice.

**Comment by vbgl** [comment]
@GrahamcOfBorg build stog

**Review by sternenseemann — APPROVED** [review]
Updated manual would be good as well.

**Comment by vbgl** [comment]
Well spotted. Will do.

File: pkgs/build-support/ocaml/dune.nix [hunk 1]
```diff
@@ -1,8 +1,11 @@
-{ lib, stdenv, ocaml, findlib, dune_1, dune_2 }:
+{ lib, stdenv, ocaml, findlib, dune_1, dune_2, dune_3 }:
 
 { pname, version, nativeBuildInputs ? [], enableParallelBuilding ? true, ... }@args:
 
-let Dune = if args.useDune2 or true then dune_2 else dune_1; in
+let Dune =
+  let dune-version = args . duneVersion or (if args.useDune2 or true then "2" else "1"); in
```
Comment (SuperSandro2000) [inline pkgs/build-support/ocaml/dune.nix:6]:
```suggestion
  let dune-version = args.duneVersion or (if args.useDune2 or true then "2" else "1"); in
```
or am I missing something here?

**Comment by vbgl** [comment]
The manual has just been updated.

**Pull request merged**